### PR TITLE
Optionally specify name of http codecs to replace when websocket handshake completes

### LINKED ISSUE #5201: WebSocket upgrade wrecks HttpProxyHandler
Short story: `WebSocketClientHandshaker` reorganizes the Pipeline after receiving the upgrade from the server. If there is an `HttpProxyHandler` in the Pipeline, it has installed its own `HttpClientCodec` which gets wrecked by `WebSocketClientHandshaker`.

I used a handcrafted configuration which was easy to hack to see where the problem is. Since proxies don't support `ws://` I used `wss://` all along. I used Squid 3.4 as a proxy. I ran those configurations:
- Firefox and Netty's WebSocket demo server. Success.
- Firefox, Netty's WebSocket demo server, and Squid in the middle. Success.
- Netty WebSocket client, Netty's WebSocket demo server. Success.
- Netty WebSocket client, Netty's WebSocket demo server, Squid in the middle. Fails.

Using Netty's `LoggingHandler` here and here, I found that adding the `HttpProxyHandler` had unexpected side-effects. I finally dumped the `ChannelPipeline` and discovered that `#finishHandshake` reconfigures the first `HttpClientCodec` in the Pipeline, as shown by this short dump:

```
HttpClientCodec#0        <-- #finishHandshake hit this one.
ws-decoder               <-- We don't want this here.
ws-encoder               <-- We don't want this here.
Http11ProxyHandler#0
SslHandler#0
HttpClientCodec#1
                         <-- ws-xxcoder should appear here.
DownendChannelHandler#0  <-- My own stuff.
```

I'm using Netty-4.1.0.CR7 and Java 1.8.0_66.

I suggest to add a method to the `WebSocketClientHandshakerFactory` for passing a pair of method objects selecting the `HttpClientCodec` or the `HttpRequestEncoder` to modify in `WebSocketClientHandshaker#finishHandshake`. Please tell me if it makes sense.

I'll be happy to write a fix, including automated tests.

### PR DESCRIPTION
**Pull request by jbaldassari**
Motivation:

As explained in #5201, if `WebSocketClientProtocolHandler` is used in conjunction with `HttpProxyHandler`, the connection fails because `WebSocketClientProtocolHandler` removes the wrong HTTP client codec when the websocket handshake completes.  It should be removing the codec associated with the websocket handler, but instead it ends up removing the codec inserted by the HTTP proxy handler, and it then inserts the websocket codec in the wrong place in the pipeline.

Modification:

This change provides two ways to work around the incompatibility between `WebSocketClientProtocolHandler` and `HttpProxyHandler`:

1. It is now possible to pass the names of the HTTP request encoder and HTTP response decoder that should be replaced to the `WebSocketClientHandshaker` or `WebSocketClientProtocolHandler`.
2. Alternatively, a subclass of `WebSocketClientProtocolHandler` can override the `findHttpRequestEncoderToReplace(ChannelPipeline)` and `findHttpResponseDecoderToReplace(ChannelPipeline)` methods to use custom logic to select the appropriate codecs from the pipeline.

Below is an example showing how the first method can be used to give `WebSocketClientHandshaker` a hint about the correct codec to replace:

```
    @Override
    public void initChannel(final SocketChannel ch) throws Exception {
        ch.pipeline().addFirst(new HttpProxyHandler(proxyAddress));
        final String httpClientCodecName = "websocketClientCodec";
        ch.pipeline().addLast(httpClientCodecName, new HttpClientCodec()); // <- specify a name for this HttpClientCodec
        ch.pipeline().addLast(new HttpObjectAggregator(MAX_HTTP_CONTENT_LENGTH_BYTES));
        ch.pipeline().addLast(
            new WebSocketClientProtocolHandler(
                webSocketURL, version, subprotocol, allowExtensions, customHeaders, maxFramePayloadLength,
                handleCloseFrames, performMasking, allowMaskMismatch, handshakeTimeoutMillis,
                httpClientCodecName, httpClientCodecName),  // <- pass the name of the HttpClientCodec here
            new WebsocketTunnelFrameCodec(),
            // Your handler here
        );
    }
```

The key in that example is that we add an `HttpClientCodec` with a specific name, and then we pass that same name to the `WebSocketClientProtocolHandler`.  In this case the name is the same for the request encoder and response decoder because we're using a single `HttpClientCodec`, but if separate encoders and decoders were used, the names can be passed individually.

If no codec names are passed to the handshaker/handler, we fall back to the original behavior of just trying to find the first codec in the pipeline, so this is backwards-compatible with current behavior.

Result:

Fixes #5201

### CODE REVIEW

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by jbaldassari** [comment]
@normanmaurer  those build failures were caused by issues resolving dependencies it seems.  Can we kick off the builds again?

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java [hunk 1]
```diff
@@ -579,6 +612,90 @@ protected String upgradeUrl(URI wsURL) {
...
+     * @throws IllegalStateException if the http response decoder could be found in the pipeline.
+     */
+    protected ChannelHandler findHttpResponseDecoderToReplace(ChannelPipeline pipeline) throws IllegalStateException {
+        // If a name was specified for the http codec, require that a handler with that name exists:
+        if (httpResponseDecoderName != null) {
+            final ChannelHandler decoder = pipeline.get(httpResponseDecoderName);
+            if (decoder == null) {
+                throw new IllegalStateException("No http response decoder found with name \"" +
+                    httpResponseDecoderName + "\"");
+            } else if (!(decoder instanceof HttpResponseDecoder || decoder instanceof HttpClientCodec)) {
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java]: nit: you can remove the `else` here as you throw in the `if`.

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker13.java [hunk 2]
```diff
@@ -167,12 +167,54 @@ public WebSocketClientHandshaker13(URI webSocketURL, WebSocketVersion version, S
...
+     *            Close the connection if it was not closed by the server after timeout specified.
+     * @param  absoluteUpgradeUrl
+     *            Use an absolute url for the Upgrade request, typically when connecting through an HTTP proxy over
+     *            clear HTTP
+     * @param httpRequestEncoderName
+     *            The name of the HTTP request encoder to be replaced when the handshake is complete
+     * @param httpResponseDecoderName
+     *            The name of the HTTP response decoder to be replaced when the handshake is complete
+     */
+    public WebSocketClientHandshaker13(URI webSocketURL, WebSocketVersion version, String subprotocol,
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker13.java]: should we just make this package-private for now ?

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker08.java [hunk 3]
```diff
@@ -166,12 +166,54 @@ public WebSocketClientHandshaker08(URI webSocketURL, WebSocketVersion version, S
...
+     *            Close the connection if it was not closed by the server after timeout specified.
+     * @param  absoluteUpgradeUrl
+     *            Use an absolute url for the Upgrade request, typically when connecting through an HTTP proxy over
+     *            clear HTTP
+     * @param httpRequestEncoderName
+     *            The name of the HTTP request encoder to be replaced when the handshake is complete
+     * @param httpResponseDecoderName
+     *            The name of the HTTP response decoder to be replaced when the handshake is complete
+     */
+    public WebSocketClientHandshaker08(URI webSocketURL, WebSocketVersion version, String subprotocol,
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker08.java]: should we just make this package-private for now ?

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker07.java [hunk 4]
```diff
@@ -164,12 +164,54 @@ public WebSocketClientHandshaker07(URI webSocketURL, WebSocketVersion version, S
...
+     *            Close the connection if it was not closed by the server after timeout specified.
+     * @param  absoluteUpgradeUrl
+     *            Use an absolute url for the Upgrade request, typically when connecting through an HTTP proxy over
+     *            clear HTTP
+     * @param httpRequestEncoderName
+     *            The name of the HTTP request encoder to be replaced when the handshake is complete
+     * @param httpResponseDecoderName
+     *            The name of the HTTP response decoder to be replaced when the handshake is complete
+     */
+    public WebSocketClientHandshaker07(URI webSocketURL, WebSocketVersion version, String subprotocol,
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker07.java]: should we just make this package-private for now ?

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker00.java [hunk 5]
```diff
@@ -114,8 +117,40 @@ public WebSocketClientHandshaker00(URI webSocketURL, WebSocketVersion version, S
...
+     *            Close the connection if it was not closed by the server after timeout specified
+     * @param  absoluteUpgradeUrl
+     *            Use an absolute url for the Upgrade request, typically when connecting through an HTTP proxy over
+     *            clear HTTP
+     * @param httpRequestEncoderName
+     *            The name of the HTTP request encoder to be replaced when the handshake is complete
+     * @param httpResponseDecoderName
+     *            The name of the HTTP response decoder to be replaced when the handshake is complete
+     */
+    public WebSocketClientHandshaker00(URI webSocketURL, WebSocketVersion version, String subprotocol,
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker00.java]: should we just make this package-private for now ?

File: codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java [hunk 6]
```diff
@@ -238,6 +246,63 @@ protected void channelRead0(ChannelHandlerContext ctx, Object msg) throws Except
...
+                    serverReceivedHandshake = true;
+                    // immediately send a message to the client on connect
+                    ctx.writeAndFlush(new TextWebSocketFrame("abc"));
+                } else if (evt instanceof WebSocketServerProtocolHandler.HandshakeComplete) {
+                    serverHandshakeComplete = (WebSocketServerProtocolHandler.HandshakeComplete) evt;
+                }
+            }
+            @Override
+            protected void channelRead0(ChannelHandlerContext ctx, Object msg) throws Exception {
+            }
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java]: add a comment that he method is empty by design.

File: codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java [hunk 7]
```diff
@@ -238,6 +246,63 @@ protected void channelRead0(ChannelHandlerContext ctx, Object msg) throws Except
...
+            public void userEventTriggered(ChannelHandlerContext ctx, Object evt) {
+                if (evt == ServerHandshakeStateEvent.HANDSHAKE_COMPLETE) {
+                    serverReceivedHandshake = true;
+                    // immediately send a message to the client on connect
+                    ctx.writeAndFlush(new TextWebSocketFrame("abc"));
+                } else if (evt instanceof WebSocketServerProtocolHandler.HandshakeComplete) {
+                    serverHandshakeComplete = (WebSocketServerProtocolHandler.HandshakeComplete) evt;
+                }
+            }
+            @Override
```
Comment (normanmaurer) [inline codec-http/src/test/java/io/netty/handler/codec/http/websocketx/WebSocketHandshakeHandOverTest.java]: nit: add empty line above

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker00.java (hunk 5 above)
Comment (jbaldassari) [reply codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker00.java]: Sure, so the idea is to not make it public until a later time when we're reasonably certain that this interface is stable?  That sounds reasonable to me, as long as there is _some_ public way to pass these parameters along.

**Comment by jbaldassari** [comment]
Thanks for the review @normanmaurer !  Just pushed up changes that I believe will address your feedback.

**Comment by normanmaurer** [comment]
@netty-bot test this please

**Comment by normanmaurer** [comment]
@slandelle ping

**Comment by slandelle** [comment]
@jbaldassari Could you please edit your commit message to follow rules: https://netty.io/wiki/writing-a-commit-message.html

And could you also please describe precisely in this commit message the current situation: how you initialize the pipeline, the state at the different steps with the issue and the correct/expected pipeline state with your fix?

**Comment by jbaldassari** [comment]
Thank you @slandelle .  I rewrote my commits into a single commit (https://github.com/netty/netty/pull/9564/commits/528d273b7c3b6ab5c7a2b6dd5b33d4cddf2e4b77) following the guidelines you mentioned.  Please let me know if there are any other changes you'd like me to make.

**Comment by normanmaurer** [comment]
@slandelle PTAL again

**Comment by normanmaurer** [comment]
@slandelle ping

**Comment by slandelle** [comment]
@jbaldassari Is there's really no way to automatically detect there are multiple HTTP client codecs in the pipeline and delete the correct one?

**Comment by normanmaurer** [comment]
@slandelle unfortunately I was not able to come up with a good way without hacks and breaking the aPI :/

**Comment by jbaldassari** [comment]
Updated to resolve conflicts

**Comment by normanmaurer** [comment]
@netty-bot test this please

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java [hunk 8]
```diff
@@ -143,13 +148,55 @@ protected WebSocketClientHandshaker(URI uri, WebSocketVersion version, String su
...
+     *            If not specified, the first {@link HttpRequestEncoder} or {@link HttpClientCodec} found in the
+     *            pipeline will be replaced.
+     * @param httpResponseDecoderName
+     *            After the handshake is complete, the {@link HttpResponseDecoder} or {@link HttpClientCodec}
+     *            in the pipeline needs to be replaced with a websocket decoder.  If httpResponseDecoderName is present,
+     *            the handshaker assumes that this is the name of the http response decoder that should be replaced.
+     *            If not specified, the first {@link HttpResponseDecoder} or {@link HttpClientCodec} found in the
+     *            pipeline will be replaced
+     */
+    protected WebSocketClientHandshaker(URI uri, WebSocketVersion version, String subprotocol,
```
Comment (slandelle) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java]: IMHO, we should not add yet another protected constructor (hence an API) and first introduce a Builder and deprecate the existing constructors.

**Comment by slandelle** [comment]
I'd have to test on my side and compare with how we do things in Gatling (we don't use WebSocketClientProtocolHandler atm). But still, I think we should introduce a Builder first.

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker08.java [hunk 9]
```diff
@@ -166,12 +166,54 @@ public WebSocketClientHandshaker08(URI webSocketURL, WebSocketVersion version, S
      *            Use an absolute url for the Upgrade request, typically when connecting through an HTTP proxy over
      *            clear HTTP
      */
+    public WebSocketClientHandshaker08(URI webSocketURL, WebSocketVersion version, String subprotocol,
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker08.java]: package-private ?

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java (hunk 8 above)
Comment (normanmaurer) [reply codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker.java]: I think it is ok to add this as long as we make it package-private

File: codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker13.java [hunk 10]
```diff
@@ -167,12 +167,54 @@ public WebSocketClientHandshaker13(URI webSocketURL, WebSocketVersion version, S
      *            Use an absolute url for the Upgrade request, typically when connecting through an HTTP proxy over
      *            clear HTTP
      */
+    public WebSocketClientHandshaker13(URI webSocketURL, WebSocketVersion version, String subprotocol,
```
Comment (normanmaurer) [inline codec-http/src/main/java/io/netty/handler/codec/http/websocketx/WebSocketClientHandshaker13.java]: package-private ?

**Comment by normanmaurer** [comment]
After thinking more about all of this I realised there is a very simple workaround which doesn't need all these changes. 

PTAL:
https://github.com/netty/netty/pull/10103

**Comment by normanmaurer** [comment]
I think we can close this now as #10103 fixes the problem

**Comment by normanmaurer** [comment]
@jbaldassari thanks a lot for the work!

**Pull request closed without merging**